File: src/main/java/org/dbtools/query/shared/QueryBuilder.java
package org.dbtools.query.shared;

import java.util.ArrayList;
import java.util.List;

@SuppressWarnings("unused")
public abstract class QueryBuilder {
    public abstract String formatLikeClause(String field, String value);

    public abstract String formatIgnoreCaseLikeClause(String field, String value);

    public abstract Object formatValue(Object value);

    public abstract String getQueryParameter();

    public abstract String buildQuery();

    public static String[] toSelectionArgs(Object... args) {
        List<String> selectionArgs = new ArrayList<String>(args.length);
        for (Object o : args) {
            if (o instanceof List) {
                for (Object p : (List) o) {
                    selectionArgs.add(String.valueOf(p));
                }
            } else {
                selectionArgs.add(String.valueOf(o));
            }
        }
        return selectionArgs.toArray(new String[selectionArgs.size()]);
    }
}

File: dbtools-query/src/commonMain/kotlin/org/dbtools/query/shared/QueryBuilder.kt
package org.dbtools.query.shared

abstract class QueryBuilder {
    abstract fun formatLikeClause(field: String, value: String): String

    abstract fun formatIgnoreCaseLikeClause(field: String, value: String): String

    abstract fun formatValue(value: Any?): Any?

    abstract val queryParameter: String

    abstract fun buildQuery(): String

    companion object {
        fun toSelectionArgs(vararg args: Any): Array<String> {
            val selectionArgs: MutableList<String> = ArrayList(args.size)
            for (o in args) {
                if (o is List<*>) {
                    for (p in o) {
                        selectionArgs.add(p.toString())
                    }
                } else {
                    selectionArgs.add(o.toString())
                }
            }
            return selectionArgs.toTypedArray<String>()
        }
    }
}
